strapi/strapi
# Online license verification

### What does it do?

It introduces an online verification of the license, which consists of the following:
- Pretty early during the startup - whenever the `strapi.EE` is read, the local license is verified to ensure that Strapi generated it, then the project is optimistically considered licensed.
This is needed as the license state is used to determine which admin panel version to build.
- Later, if applicable, during "bootstrap" the license is updated online to get the latest data. The data is stored in the database as a potential fallback and a way to avoid too many unnecessary requests for horizontally scaled projects.
Also, the content of the license is checked to enable the right features and verify the expiration date. At this step, the project can switch back to the community edition.

@alexandrebodin introduced a wrapper and a middleware to ensure that EE features could not be consumed on an unlicensed project even if the EE admin panel was used.

### Why is it needed?

The offline system that we are currently using has its limitations.
For instance, a user has to restart the app **after updating the project's license** whenever the feature he/she has access to changes since the content of the license would be outdated.

### How to test it?

You can make sure that you're project is licensed when using a valid license and that only one outbound call should be made, even when starting your project across multiple processes.

It should be working on every supported database, SQLite doesn't support the `FOR UPDATE` syntax, so when multiple processes are started simultaneously, multiple requests will be sent.
It's not a big deal, as SQLite should be used in production rarely.

### What's missing?

- [x] Deploy the changes required on the license registry
- [x] Make sure the locking mechanism leveraged in the database for the license works as expected. I noticed something with SQLite.
- [x] Schedule a cron job that will be responsible for periodically updating the license, as well as verifying its content to update the project if needed.

This PR was co-authored with @alexandrebodin.
